borg/archiver.py

```diff
@@ -355,9 +355,10 @@ def do_extract(self, args):
         dirs = []
         partial_extract = not matcher.empty() or strip_components
         items = {} if partial_extract else None
-        for item in archive.iter_items(preload=True):
+        for item in archive.iter_items(preload=True,
+                filter=lambda item: item.get(b'link', True) or matcher.match(item[b'path'])):
             orig_path = item[b'path']
-            if partial_extract and stat.S_ISREG(item[b'mode']):
+            if partial_extract and stat.S_ISREG(item[b'mode']) and item.get(b'link', True):
```

Comment:
Yes, with all the drawbacks mentioned above (a lot higher memory usage, broken/useless preloading for remote repositories, since all items are marked for preloading as if you were extracting the entire archive).